File: src/golang/p100to199/147.对链表进行插入排序.go
package p100to199

import "sort"

/*
 * @lc app=leetcode.cn id=147 lang=golang
 *
 * [147] 对链表进行插入排序
 *
 * https://leetcode-cn.com/problems/insertion-sort-list/description/
 *
 * algorithms
 * Medium (65.52%)
 * Likes:    245
 * Dislikes: 0
 * Total Accepted:    47.2K
 * Total Submissions: 71.5K
 * Testcase Example:  '[4,2,1,3]'
 *
 * 对链表进行插入排序。
 *
 *
 * 插入排序的动画演示如上。从第一个元素开始，该链表可以被认为已经部分排序（用黑色表示）。
 * 每次迭代时，从输入数据中移除一个元素（用红色表示），并原地将其插入到已排好序的链表中。
 *
 *
 *
 * 插入排序算法：
 *
 *
 * 插入排序是迭代的，每次只移动一个元素，直到所有元素可以形成一个有序的输出列表。
 * 每次迭代中，插入排序只从输入数据中移除一个待排序的元素，找到它在序列中适当的位置，并将其插入。
 * 重复直到所有输入数据插入完为止。
 *
 *
 *
 *
 * 示例 1：
 *
 * 输入: 4->2->1->3
 * 输出: 1->2->3->4
 *
 *
 * 示例 2：
 *
 * 输入: -1->5->3->4->0
 * 输出: -1->0->3->4->5
 *
 *
 */

/**
 * @File    :   147.对链表进行插入排序.go
 * @Time    :   2020/11/20 09:32:00
 * @Author  :   wylu
 * @Version :   1.0
 * @Contact :   15wylu@gmail.com
 * @License :   Copyright © 2020, wylu-CHINA-SHENZHEN. All rights reserved.
 * @Desc    :
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func insertionSortList(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}

	assist := []*ListNode{}
	for head != nil {
		assist = append(assist, head)
		head = head.Next
	}

	sort.Slice(assist, func(i, j int) bool {
		return assist[i].Val < assist[j].Val
	})

	for i := 1; i < len(assist); i++ {
		assist[i-1].Next = assist[i]
	}
	assist[len(assist)-1].Next = nil

	return assist[0]
}

// @lc code=end

File: src/python/p100to199/147.对链表进行插入排序.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
@File    :   147.对链表进行插入排序.py
@Time    :   2020/11/20 09:15:47
@Author  :   wylu
@Version :   1.0
@Contact :   15wylu@gmail.com
@License :   Copyright © 2020, wylu-CHINA-SHENZHEN. All rights reserved.
@Desc    :
"""

#
# @lc app=leetcode.cn id=147 lang=python3
#
# [147] 对链表进行插入排序
#
# https://leetcode-cn.com/problems/insertion-sort-list/description/
#
# algorithms
# Medium (65.52%)
# Likes:    245
# Dislikes: 0
# Total Accepted:    47.2K
# Total Submissions: 71.5K
# Testcase Example:  '[4,2,1,3]'
#
# 对链表进行插入排序。
#
#
# 插入排序的动画演示如上。从第一个元素开始，该链表可以被认为已经部分排序（用黑色表示）。
# 每次迭代时，从输入数据中移除一个元素（用红色表示），并原地将其插入到已排好序的链表中。
#
#
#
# 插入排序算法：
#
#
# 插入排序是迭代的，每次只移动一个元素，直到所有元素可以形成一个有序的输出列表。
# 每次迭代中，插入排序只从输入数据中移除一个待排序的元素，找到它在序列中适当的位置，并将其插入。
# 重复直到所有输入数据插入完为止。
#
#
#
#
# 示例 1：
#
# 输入: 4->2->1->3
# 输出: 1->2->3->4
#
#
# 示例 2：
#
# 输入: -1->5->3->4->0
# 输出: -1->0->3->4->5
#
#
#


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


# @lc code=start
class Solution:
    def insertionSortList(self, head: ListNode) -> ListNode:
        if not head:
            return None

        assist = []
        while head:
            assist.append(head)
            head = head.next

        assist.sort(key=lambda node: node.val)
        for i in range(1, len(assist)):
            assist[i - 1].next = assist[i]
        assist[-1].next = None

        return assist[0]


# @lc code=end
